Allow derivatives to be referenced in figures (#3046)
modelica/ModelicaSpecification: chapters/annotations.tex

end Curve;
\end{lstlisting}

The mandatory \lstinline!x! and \lstinline!y! expressions are restricted to be result references in the form of \lstinline[language=grammar]!result-reference! in the grammar (\cref{expressions1}), referring to a scalar variable (or a derivative thereof) or \lstinline!time!.
It is an error if \lstinline!x! or \lstinline!y! does not designate a scalar result.
If \lstinline!x! or \lstinline!y! is a derivative, \lstinline!der(v, $n$)!, then $n$ must not exceed the maximum amount of differentiation applied to \lstinline!v! in the model.
When the \lstinline!unit! of an \lstinline!Axis! is non-empty, it is an error if the unit of the corresponding \lstinline!x! or \lstinline!y! expression (i.e., a variable's \lstinline!unit!, or second for \lstinline!time!) is incompatible with the axis unit.

The set of available result variable and derivative trajectories in a simulation result is tool-specific, but it is recommended to include at least the ones being used in figures.

> It is not standardized how a tool should handle missing results for valid result references.

Author: Personally, I don't like having this kind of statements about what isn't standardized in the specification.  It was included as a precaution against the otherwise inevitable comments about what a tool is supposed to do with any derivatives it didn't store in the result.  (Just trying to bring some extra attention to this sentence in case others have similar mixed feelings.)

Reviewer: I would prefer to not add the first sentence. The reason is that it suggests that a tool must use the figure-annotations when generating the result file (so it requires using figure annotation as part of translation/simulation). In Dymola you can set up what variables to store in different ways and it wouldn't be clear how figure-annotations should override that - especially as users also want to modify that for various reasons.

For the second sentence I would prefer that missing results generate a diagnostic in some way, as currently it suggests that a tool could plot a curve with zero value or even some other variable.

That also provides a solution for the first issue, as tools may suggest how to include that variable in the result in the diagnostics (including having a button to change the setup) - without having to always include those variables.

Author: > That also provides a solution for the first issue, as tools may suggest how to include that variable in the result in the diagnostics (including having a button to change the setup) - without having to always include those variables.

The point of making a recommendation was to not make a forcing statement about what tools should store, and it seems important to not remove the tools' freedom to handle this in different ways.  @otronarp, maybe you too find making recommendations for what to be stored too intrusive?

Since this question is much bigger than just figures, we should also drag more people's attention to this question.  To name a few: @gkurzbach, @eshmoylova, @sjoelund, @axelmartenssonmodelon.  The upcoming web meeting would also be a good forum for discussing this.

Reviewer 2: > The point of making a recommendation was to not make a forcing statement about what tools should store, and it seems important to not remove the tools' freedom to handle this in different ways. @otronarp, maybe you too find making recommendations for what to be stored too intrusive?

Since it was written as a recommendation I felt free to ignore it, so I saw no harm in having it. On the other hand what's the point of having the recommendation if it's going to be ignored.

Reviewer: I prefer to not have recommendations that are normally ignored, and I also prefer somewhat stronger suggestions - so that missing results generate some form of warning.

> The set of available result variable and derivative trajectories in a simulation result is tool-specific, but it is recommended to include at least the ones being used in figures.
> It is not standardized how a tool should handle missing results for valid result references.

Reviewer: ```suggestion
Missing results for valid result references, should generate some form of diagnostics.
```
As indicated I don't like the ignored recommendation of line 196 and I want to specify allowed behavior in line 197. We could add that tools may suggest how to add the missing results, but it seems almost redundant.

Reviewer: Agreement (except for comma and "some form").

Author: Done.  OK?

\begin{nonnormative}
While the syntax for referring to a second order derivative is \lstinline!der(v, 2)!, the appearance is left for tools to decide.

> For example, a tool might choose to present this as \lstinline!der(der(v))! or even $v''$.

Reviewer: Personally I would prefer `\ddot{v}` over `$v''$` to denote the time-derivative, although it is less important as it is non-normative. See https://en.wikipedia.org/wiki/Notation_for_differentiation
```suggestion
For example, a tool might choose to present this as \lstinline!der(der(v))! or even $\ddot{v}$.
```

Reviewer 3: This looks really a matter of taste. BTW I'm having endless discussions with my friends from thermodynamics, who insist on using $\dot{m}$ for flow rate and $\dot{Q}$ for thermal power. It's hopeless

Reviewer: To me dot is for time-derivatives, and prime for more general derivatives.
(And flow rate and thermal power aren't really derivatives.)

Reviewer 3: To me too. But I couldn't convince them. And they also use Modelica. Sometimes life's complicated...

Author: Personally, I like to reserve `$\dot{x}$` for something like a function formal parameter meant for evaluation at `$x'(t)$`, where `$x'$` is the derivative of `$x$`.  In the context of simulation results, each notation is meaningful:
- `$x'$` corresponds more closely to `der(x)` being thought of as the differentiation operator applied to `x`.
- `$\dot{x}$` would emphasis that this is should not be seen as (plotting) an expression of `x`, but as (plotting) the independent result "variable" `$\dot{x}$`.

While the interpretations above actually speak in favor of `$\dot{x}$`, it was a mistake from the beginning to use math style for the variable `v`.  Presenting variable names in math style seems like a stylistic mess to me, so it would have been better to have this discussion around `$\text{\lstinline{v}}'$`.  If we then think of a longer variable name, such as `aimcM.rotorCage.electroMagneticConverter.singlePhaseElectroMagneticConverter[1].Phi.re`, then I think it becomes clear that the `$\dot{…}$ approach has problems.

Can't we just drop the $v''$ variant from the text?

Reviewer: Yes, I agree that dropping `$v''$` is probably best.
